### DISCUSSION TITLE
Async file write or with shutil save file from buffer, which one better/corret way?

### DISCUSSION THREAD

**Discussion by cihatertem** [fastapi/fastapi#9618] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
async def save_uploaded_image(image: UploadFile) -> str:
    """Saves uploaded image and return new filename."""
    new_filename = sign_filename_with_uuid(image.filename)

    path = os.path.join("/vol", "static", "images", new_filename)

    image_content = await image.read()

    async with aiofiles.open(path, "w+b") as f:
        await f.write(image_content)

    return new_filename

vs

async def save_uploaded_image(image: UploadFile) -> str:
    """Saves uploaded image and return new filename."""
    new_filename = sign_filename_with_uuid(image.filename)

    path = os.path.join("/vol", "static", "images", new_filename)

    with open(path, "w+b") as buffer:
        shutil.copyfileobj(image.file, buffer)

    return new_filename
```

### Description

I wonder which way is better or correct? Both ways save the image without problem but I dont know which one is better for performance or other things.

Thanks,
regards.

### Operating System

Linux

### Operating System Details

Ubuntu 22.04.2

### FastAPI Version

0.96.0

### Python Version

3.10.6

**Comment by jgould22** [comment, marked as answer]
Shutil will block the event loop as it is doing synchronous file access. You do not want to do that as it will degrade performance. 

`aiofiles` if I recall correctly is essentially a wrapper around the file operations that just spawns synchronous calls into a thread pool. 

Another library `aiofile` (note the lack of `s`) seeks to use actual asynchronous operating system operations backed by libaio(which are relatively 'new') though it has a far slimmer API.

Essentially you should choose one of the later 2, likely the `aiofiles` which you have already used as it provides the most feature complete API and will not block the event loop.
